New package simplicial modules

kellerlv opened the pull request:
A package for computing simplicial modules and the Dold-Kan correspondence, by Michael DeBellevue and Keller VandeBogert

kellerlv committed:
New package simplicial modules

kellerlv committed:
Fixed typos

d-torrance reviewed (changes requested):
Thanks for the contribution!

Currently, test 5 is failing with the following error:

```m2
i13 : 	    assert isSimplicial C2
stdio:15:31:(3): error: no method found for applying isSimplicial to:
     argument   :   1      1      1      1      1 . (of class SimplicialModule)
                   S  <-- S  <-- S  <-- S  <-- S  .
                                                  .
                   0      1      2      3      4  .
```

kellerlv committed:
Bug fix in test 5

(forgot to update old function name)

d-torrance commented:
The tests are failing on one of the builds due to running out of memory:

```m2
-- capturing check(47, "SimplicialModules") Insufficient memory for the allocation
```

Is there any way of making this example smaller?  There are a few others that take a while:

```m2
-- capturing check(9, "SimplicialModules")   -- 6.0101s elapsed
...
-- capturing check(26, "SimplicialModules")  -- 69.118s elapsed
-- capturing check(27, "SimplicialModules")  -- 22.6293s elapsed
```

kellerlv commented:
> The tests are failing on one of the builds due to running out of memory:
> 
> ```m2
> -- capturing check(47, "SimplicialModules") Insufficient memory for the allocation
> ```
> 
> Is there any way of making this example smaller? There are a few others that take a while:
> 
> ```m2
> -- capturing check(9, "SimplicialModules")   -- 6.0101s elapsed
> ...
> -- capturing check(26, "SimplicialModules")  -- 69.118s elapsed
> -- capturing check(27, "SimplicialModules")  -- 22.6293s elapsed
> ```

Yes -- I should have pruned this one down. I had to fix a similar example in the documentation that was causing memory issues so I should have done a similar tweak here.

kellerlv committed:
Pruned down tests

Should now run without taking too much memory

d-torrance commented on M2/Macaulay2/packages/SimplicialModules.m2:
```diff
@@ -0,0 +1,208 @@
+     Headline => "methods for working in the category of simplicial modules",
+     Keywords => {"Homological Algebra", "Commutative Algebra"},
+     PackageExports => {"Complexes", "SchurFunctors"},
+     PackageImports => {"Complexes"}
```
Having Complexes in both PackageImports and PackageExports is redundant

kellerlv replied:
What's the preferred threshold for test lengths? These tests took about half as long as those times on my computer, but I acknowledge 35 seconds is quite long. One issue with this is that simplicial objects are generally very big and computationally cumbersome, so it's hard to build nontrivial examples that also run quickly. I kind of knew the example sizes would be an issue here but wasn't sure what was considered acceptable.

mahrud replied:
What's the main bottleneck? Have you profiled the test?

kellerlv replied:
No, I haven't profiled it mostly because I don't know how. For all of the slow tests it generally stems not from this package's constructors themselves, but the general slowness of applying Schur functors/exterior powers to large matrices (which are functions from other packages).

d-torrance commented on M2/Macaulay2/packages/SimplicialModules/Normalization.m2:
```diff
@@ -0,0 +1,311 @@
+ needsPackage "SchurFunctors"
+ needsPackage "Complexes"
```
These calls to `needsPackage` are unnecessary since the packages are in `PackageExports`

d-torrance commented on M2/Macaulay2/packages/SimplicialModules/Normalization.m2:
```diff
@@ -0,0 +1,311 @@
+schurMap(List,SimplicialModule) := SimplicialModule => opts -> (lambda,S) -> (tdeg := topDegree S;
+    --C := S.complex;
+    L := hashTable for i to tdeg list i => schurModule(lambda,combineSFactors(S,i));
+    H1 := hashTable for i in keys (S.dd.map) list i => schur(lambda,((S.dd)_i));
```
Using `applyValues` here would be a bit cleaner here since we wouldn't need to create an intermediate list

d-torrance commented on M2/Macaulay2/packages/SimplicialModules/Normalization.m2:
```diff
@@ -0,0 +1,311 @@
+    --C := S.complex;
+    L := hashTable for i to tdeg list i => schurModule(lambda,combineSFactors(S,i));
+    H1 := hashTable for i in keys (S.dd.map) list i => schur(lambda,((S.dd)_i));
+    if opts.Degeneracy==true then H2 := hashTable for i in keys (S.ss.map) list i => schur(lambda,((S.ss)_i));
```
Comparing to `true` here is unnecessary since `opts.Degeneracy` is already a boolean

mahrud replied:
(Although since it's an option, `=== true` would make sure it's not null also.

d-torrance commented on M2/Macaulay2/packages/SimplicialModules/Normalization.m2:
```diff
@@ -0,0 +1,311 @@
+schurMap(List,SimplicialModuleMap) := SimplicialModuleMap => opts -> (lambda,phi) -> (
+    S1 := source phi;
+    S2 := target phi;
+    if instance((keys phi.map)#0,Sequence) then error "Expected SimplicialModuleMap to have singly graded indices";
```
Usually error message begin with a lowercase letter

d-torrance commented on M2/Macaulay2/packages/SimplicialModules/Normalization.m2:
```diff
@@ -0,0 +1,311 @@
+    map(schurMap(lambda,S2),schurMap(lambda,S1),new HashTable from for i to max(topDegree S1,topDegree S2) list i => schur(lambda,phi_i),Degree => degree phi)
+    )
+
+schurMap(List,Complex) := Complex => opts -> (lambda,C) -> (S := if not(opts.TopDegree === null) then simplicialModule(C,opts.TopDegree)
```
You can use `=!=` here to negate `===`

d-torrance reviewed (changes requested):
I've added a few additional comments on the code.  Also, some of the tests are still pretty lengthy:

```m2
 -- capturing check(26, "SimplicialModules") -- 35.9748s elapsed
 -- capturing check(27, "SimplicialModules") -- 18.2174s elapsed
...
-- capturing check(47, "SimplicialModules") -- 7.68881s elapsed
...
 -- capturing check(49, "SimplicialModules") -- 8.55586s elapsed
 -- capturing check(50, "SimplicialModules") -- 20.7422s elapsed
```

kellerlv committed:
Major overhaul of SimplicialModules: efficiency, correctness, and documentation

- Optimized face map construction with fast sparse exterior power and
  Schur functor specializations, AMatrix memoization, and direct
  block-matrix construction for the i=0 face map (faceMap0Direct)
- Fixed faceMap0Direct source/target consistency with other face maps
  to prevent invalid complexes in naiveNorm
- Fixed isShortExactSequence to check exactness degree-by-degree,
  avoiding false negatives from different generator representations
- Fixed prune to call forgetComplex first, preventing .complex cache
  issues with non-free modules
- Fixed map equality, composition, and RingMap operations for
  quotient rings and non-free modules
- Added forgetComplex with RememberSummands option for robust
  directSum component preservation
- Cleaned up documentation: added option stubs, fixed cross-references,
  added HomePage metadata
- Added 9 new regression tests covering all bug fixes

kellerlv committed:
Address reviewer feedback and fix CI doc example failures

- Fix faceMap0Direct: use flat directSum instead of fold(directSum,...)
  to preserve component indexing for _[k]^[j] array notation
- Fix SimplicialModule Array: remove forgetComplex so shifted modules
  retain .complex cache for normalize/randomComplexMap compatibility
- Apply d-torrance review: remove redundant PackageImports, remove
  unnecessary needsPackage calls, lowercase error messages, use === true
  for option checks, use =!= null instead of not(... === null)

kellerlv committed:
Merge branch 'development' into SimplicialModules

d-torrance commented on M2/Macaulay2/packages/SimplicialModules.m2:
```diff
@@ -0,0 +1,208 @@
+------------------------------------------------------------------------------
+
+load "./SimplicialModules/SimplicialMapUtilities_Modified.m2"
+load "./SimplicialModules/SimplicialModule.m2"
```
We're loading `SimplicialModule.m2`, but there's also a `SimplicialModule_Modified.m2` with much of the same content.  That file should be deleted.

Similarly, perhaps rename `SimplicialMapUtilities_Modified.m2` -> `SimplicialMapUtilities.m2`.

kellerlv committed:
Remove unused _Modified.m2 file and rename SimplicialMapUtilities

kellerlv commented:
Is this last failure on my end? I looked at where the package failures occurred and they seem to be unrelated to this package?

d-torrance commented:
> Is this last failure on my end? I looked at where the package failures occurred and they seem to be unrelated to this package?

Yup -- completely unrelated.  The test failure is a random one we see every once in a while.  Nothing to worry about on your end!

d-torrance commented on M2/Macaulay2/packages/SimplicialModules.m2:
```diff
@@ -0,0 +1,208 @@
+    "SimplicialModules",
+     AuxiliaryFiles => true,
+     Version => "0.1",
+     Date => "September 27, 2023",
```
This date should probably be updated :)

d-torrance commented on M2/Macaulay2/packages/SimplicialModules.m2:
```diff
@@ -0,0 +1,208 @@
+     Date => "September 27, 2023",
+     Authors => {
+	{Name => "Keller VandeBogert", Email => "kvandebo@nd.edu", HomePage => "https://sites.google.com/view/kellervandebogert/home"},
+	{Name => "Michael DeBellevue", Email => "", HomePage => ""}},
```
Does Michael have an email address and/or homepage?

d-torrance commented on M2/Macaulay2/packages/SimplicialModules.m2:
```diff
@@ -0,0 +1,208 @@
+     Headline => "methods for working in the category of simplicial modules",
+     Keywords => {"Homological Algebra", "Commutative Algebra"},
+     PackageExports => {"Complexes", "SchurFunctors"},
+     HomePage => "https://github.com/kellervandebogert/SimplicialModules"
```
This link is broken

d-torrance commented on M2/Macaulay2/packages/SimplicialModules.m2:
```diff
@@ -0,0 +1,208 @@
+    "RememberSummands",
+    "Degeneracy",
+    "TopDegree",
+    "ss",
```
What's `ss` for?  I'd argue that this is too short to export

kellerlv replied:
Actually, ss is supposed to be the analogue of the dd call for the differentials in a complex. For simplicial objects there are degeneracy maps as well, and a common notation in the literature is to use s (or sigma) to denote these maps. So it seems reasonable keep this exported since it fits with the pre-established M2 syntax for complexes, and it's a fundamental piece of the data of the objects in this package.

d-torrance commented on M2/Macaulay2/packages/SimplicialModules.m2:
```diff
@@ -0,0 +1,208 @@
+-----------------------------------------------------------------------------
+
+end--
+
```
I'd drop all this unused code after `end`.  If they haven't already been used for this, they would probably make good candidates for tests and/or examples

mahrud replied:
I'll just add my two cents to say having _something_ after the `end--` part is extremely useful for other people trying to learn what examples the authors actually care about or what sort of things they are still working on.

d-torrance replied:
Fair enough.  It's definitely a good place for more computationally intensive code that wouldn't be good fits for docs/tests.

kellerlv replied:
I'm just going to leave it then because some of them aren't easy to convert into tests/documentation, but might still be useful to other users.

d-torrance commented on M2/Macaulay2/packages/SimplicialModules/SimplicialModule.m2:
```diff
@@ -0,0 +1,1446 @@
+    -- check keys, check their types
+    if not instance(C.ring, Ring) then (
+        if debugLevel > 0 then (
+            << "-- expected 'ring C' to be a ring" << endl;
```
Minor suggestion:  this could be `printerr("expected 'ring C' to be a ring")` (and some for the other positive `debugLevel` messages)

d-torrance reviewed (changes requested):
Thanks!  Just a few more small comments

kellerlv committed:
Address reviewer feedback: package metadata, ss doc, printerr conversion

- Update package Date to April 27, 2026
- Add Michael DeBellevue's email address; drop empty HomePage field
- Remove broken package-level HomePage URL
- Add etymology blurb in the existing face/degeneracy maps doc node so
  the meaning of the exported symbol ss is discoverable
- Convert positive-debugLevel diagnostic blocks in SimplicialModule.m2
  from `<< "-- msg" << endl;` to `printerr "msg"`, matching M2 Core
  house style and routing diagnostics to stderr

kellerlv committed:
Merge remote-tracking branch 'upstream/development' into SimplicialModules

kellerlv committed:
SchurFunctors: drop "normalize" from export list

The recent Weyl-functor commit (fca120e064) added "normalize" to the
SchurFunctors export list. This conflicts with SimplicialModules, which
defines its own normalize method with options on SimplicialModule. Since
"normalize Filling" was unexported in SchurFunctors prior to that commit,
no existing user code can depend on the export, and removing it restores
compatibility between the two packages.

d-torrance commented on M2/Macaulay2/packages/SchurFunctors.m2:
```diff
@@ -20,7 +20,7 @@ newPackage(
 export{ "schur", "schurModule", "Filling",
      "straighten", "printSchurModuleElement", "schurModulesMap", "augmentFilling",
-     "isStandard", "standardTableaux", "normalize",
+     "isStandard", "standardTableaux",
```
This change is causing an example failure:

```m2
 -- error log:  /home/runner/work/M2/M2/M2/BUILD/build/usr-dist/common/share/doc/Macaulay2/SchurFunctors/example-output/_normalize.errors:0:1:(3):[14]: stdio:1:9:(3):[1]: error: no method for adjacent objects:
 --             normalize (of class Symbol)
 --             +-+-+
 --     SPACE   |1|2| (of class Filling)
 --             |0|1|
 --             +-+-+
 -- 
 -- input code: /tmp/M2-36108-0/0_normalize.m2:0:1
*** error: M2 exited with status code 1 -- 4.26559s elapsed
```

You could call `exportFrom("SimplicialModules", "normalize")` instead.  (Although we should probably think about adding it to `Core` as a shared symbol at some point...)

kellerlv replied:
Thanks, I didn't know about using exportFrom. This saved me since I was just about to rename every instance of normalize in this package since it was annoying me so much. We'll see if it builds correctly this time ..

kellerlv committed:
share normalize between SchurFunctors and SimplicialModules

Following d-torrance's suggestion, use exportFrom instead of dropping
the export. SchurFunctors owns the symbol with method(Options => true);
SimplicialModules picks it up via exportFrom_SchurFunctors and installs
its own methods with options inlined per install site. Docs split
accordingly.

mahrud commented on M2/Macaulay2/packages/SimplicialModules.m2:
```diff
@@ -42,6 +41,10 @@ export {"SimplicialModule",
 
+-- normalize is owned and exported by SchurFunctors; we share the symbol so
+-- this package can install its own normalize methods on simplicial modules.
+exportFrom_SchurFunctors { "normalize" }
```
I believe you want `importFrom`? Since `SchurFunctors` is in PackageExports, the symbol `normalize` is already exported.

ps: better would be to move `normalize = method(Options => true)` to `m2/shared.m2`, but we can do this later.

kellerlv replied:
Is this what's causing the current failure? I was trying to parse the failure on my own but it seems like the build didn't even get to loading the packages, so I've been at a loss for how to fix this.

mahrud replied:
No, I suspect a cosmic ray hit and run. I'll restart the build.

d-torrance replied:
*SchurFunctors* doesn't export it, though, so I think we need `exportFrom` if the goal is to export it

mahrud replied:
ps: not a cosmic ray, but I doubt it's your fault that the autotools-ubuntu build keeps failing.

kellerlv replied:
The PR before this I had unexported "normalize" since I thought that would fix the failing test, then I re-exported it and instead used exportFrom. I'll try using importFrom instead.

mahrud commented on M2/Macaulay2/packages/SchurFunctors.m2:
```diff
@@ -20,7 +20,7 @@ newPackage(
 export{ "schur", "schurModule", "Filling",
      "straighten", "printSchurModuleElement", "schurModulesMap", "augmentFilling",
-     "isStandard", "standardTableaux",
+     "isStandard", "standardTableaux", "normalize",
```
@d-torrance isn't SchurFunctors exporting `normalize` here?

d-torrance replied:
Ah -- that's the problem!

Prior to #4197, `normalize` *wasn't* exported, so *SimplicialModules* was free to re-define it.  But once that PR was merged, all of a sudden there was a conflict.  That's why this just showed up now.

So there's no reason for `importFrom` or `exportFrom`.  Since *SchurFunctors* is already in `PackageExports`, we can go ahead and install `normalize` methods to our heart's content, so long as we *don't* have `normalize = method()` in this package.

kellerlv replied:
I think I got it working. We'll see if it builds correctly this time.

kellerlv committed:
13th time's the charm

d-torrance reviewed (approved):
Thanks!
